Automattic/jetpack
# Search: migrate/add search rest API

#### Changes proposed in this Pull Request:
Add all necessary APIs for standalone and Jetpack the plugin except for the `search_auto_config` API and probably the API to deactivate expired search.

1. Returns search plan: GET jetpack/v4/search/plan

```JSON
{
  "search_subscriptions": [
    {
      "ID": "xxx",
      "user_id": "xxx",
      "blog_id": "xxx",
      "product_id": "2104",
      "expiry": "2025-05-17",
      "subscribed_date": "2021-05-17 05:34:09",
      "renew": "0",
      "auto_renew": "1",
      "ownership_id": "xxx",
      "most_recent_renew_date": "",
      "subscription_status": "active",
      "is_instant_search": true,
      "tier": "up_to_100_records",
      "product_name": "Jetpack Search"
    }
  ],
  "supports_instant_search": true,
  "supports_only_classic_search": false,
  "supports_search": true
}
```

2. Returns search module settings: GET jetpack/v4/search/settings

```JSON
{
  "module_active": true,
  "instant_search_enabled": false
}
```

3. Update search module settings: POST jetpack/v4/search/settings

Request: 
```JSON
{
  "instant_search_enabled": false,
  "module_active": true
}
```
Response:
```JSON
{
  "module_active": true,
  "instant_search_enabled": false
}
```

4. Search endpoint for private sites (private Atomic sites atm): GET jetpack/v4/search

The format is exactly the same with public search API v1.3. Something like the following,
```
aggregations: []
corrected_query: false
page_handle: false
results: [{_score: 2,…}, {_score: 1.1753112,…}, {_score: 1.1753112,…}, {_score: 1.1753112,…},…]
suggestions: []
total: 6
```

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Make sure tests pass `( cd projects/packages/search && composer phpunit )`
- Go to a site with search subscription
- Set `blog_public` to `-1` in the database
- Ensure instant search calls `http://yousite/wp-json/jetpack/v4/search` for results and results / filters show correctly.
- Test the APIs separately, I used [Rest Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) in VSCode. Requests are like the following. You'll need to replace the nonce and cookie which could be get from Chrome dev tools. If you have better ways to test, please let me know 😄 


```
GET https://jasper.au.ngrok.io/wp-json/jetpack/v4/search/plan HTTP/1.1
x-wp-nonce: xxx
cookie: yyy
```

```
GET https://jasper.au.ngrok.io/wp-json/jetpack/v4/search/settings HTTP/1.1
x-wp-nonce: xxx
cookie: yyy
```

```
POST https://jasper.au.ngrok.io/wp-json/jetpack/v4/search/settings HTTP/1.1
Content-Type: application/json
x-wp-nonce: xxx
cookie: yyy

{
  "instant_search_enabled": false,
  "module_active": true
}
```

```
GET https://jasper.au.ngrok.io/wp-json/jetpack/v4/search HTTP/1.1
x-wp-nonce: xxx
cookie: yyy
```
